turbopack/crates/turbo-tasks-backend/src/data.rs

```diff
@@ -83,6 +85,140 @@ transient_traits!(RootState);
+            count,
+            count_in_session: None,
+        })
+    }
+
+    pub fn update_session_dependent(
+        &mut self,
+        ignore_session: SessionId,
+        count: i32,
+    ) -> DirtyContainerCount {
+        self.update_count(&DirtyContainerCount {
+            count,
+            count_in_session: Some((ignore_session, 0)),
+        })
+    }
+
+    pub fn update_count(&mut self, count: &DirtyContainerCount) -> DirtyContainerCount {
+        let mut diff = DirtyContainerCount::default();
+        match (
+            self.count_in_session.as_mut(),
+            count.count_in_session.as_ref(),
+        ) {
+            (None, None) => {}
+            (Some((s, c)), None) => {
+                let d = add_with_diff(c, count.count);
+                diff.count_in_session = Some((*s, d));
+            }
+            (None, Some((s, c))) => {
+                let mut new = self.count;
+                let d = add_with_diff(&mut new, *c);
+                self.count_in_session = Some((*s, new));
+                diff.count_in_session = Some((*s, d));
+            }
+            (Some((s1, c1)), Some((s2, c2))) => match (*s1).cmp(s2) {
+                Ordering::Less => {
+                    let mut new = self.count;
+                    let d = add_with_diff(&mut new, *c2);
+                    self.count_in_session = Some((*s2, new));
+                    diff.count_in_session = Some((*s2, d));
+                }
+                Ordering::Equal => {
+                    let d = add_with_diff(c1, *c2);
+                    diff.count_in_session = Some((*s1, d));
+                }
+                Ordering::Greater => {
+                    let d = add_with_diff(c1, count.count);
+                    diff.count_in_session = Some((*s1, d));
+                }
+            },
+        }
+        let d = add_with_diff(&mut self.count, count.count);
+        diff.count = d;
+        diff
+    }
+
+    pub fn update_with_dirty_state(&mut self, dirty: &DirtyState) -> DirtyContainerCount {
+        if let Some(clean_in_session) = dirty.clean_in_session {
+            self.update_session_dependent(clean_in_session, 1)
+        } else {
+            self.update(1)
+        }
+    }
+
+    pub fn undo_update_with_dirty_state(&mut self, dirty: &DirtyState) -> DirtyContainerCount {
+        if let Some(clean_in_session) = dirty.clean_in_session {
+            self.update_session_dependent(clean_in_session, -1)
+        } else {
+            self.update(-1)
+        }
+    }
+
+    pub fn replace_dirty_state(
+        &mut self,
+        old: &DirtyState,
+        new: &DirtyState,
+    ) -> DirtyContainerCount {
+        let mut diff = self.undo_update_with_dirty_state(old);
+        diff.update_count(&self.update_with_dirty_state(new));
+        diff
+    }
+
+    pub fn is_default(&self) -> bool {
+        self.count == 0 && self.count_in_session.map(|(_, c)| c == 0).unwrap_or(true)
+    }
```

Comment:
I feel like `is_default` isn't quite right here because you're ignoring the session id.

IMO `is_default` implies that `value.is_default()` is equivalent to `value == DirtyContainerCount::default()`.

What about `is_zero`?